README.md
# Language clustering of the MSD musicXmatch dictionary 

musiXmatch contributed the lyrics of 
[237,662 tracks](https://labrosa.ee.columbia.edu/millionsong/musixmatch) 
to the [Million Song Dataset](https://labrosa.ee.columbia.edu/millionsong/). 
These lyrics are restricted to the top 5,000 words across the set, 
which contains several languages. The dataset doesn't tell in which language 
each song is written, nor to which language(s) each word might belong. 
We do not know how many languages there are in the dataset, nor their proportions 
in the dictionary. Nevertheless, it may be useful to group words and/or songs per language. 

This project uses hierarchical clustering to group words by language, 
based on word co-occurences in the dataset. The rationale for using hierarchical 
clustering over other clustering methods is that a) we do no know how many 
clusters/languages there might be in the dataset and b) we can expect 
clusters/languages to have very different word counts. 
Hierarchical clustering places every word on a same footing and allows 
for the detection of clusters of vastly varying sizes.

## Output

The clustered dictionary of the musicXmatch dataset can be found in file 
[mxm_clustered_dictionary.txt](https://raw.githubusercontent.com/belovehq/msd-languages/master/resources/public/mxm_clustered_dictionary.txt). 
A [Jupyter notebook](https://nbviewer.jupyter.org/gist/duchenne/77b54f6adb7783848edb9a6e97880c64) 
explains how the clusters were chosen.

Note that this dictionary is skewed towards the most represented languages in the dataset, 
e.g. the word "in" is classified as English despite belonging to many other languages.

The analysis is based on the musiXmatch dataset, the official lyrics collection 
for the Million Song Dataset, available at: http://labrosa.ee.columbia.edu/millionsong/musixmatch

## Computation

The analysis is written in Clojure.

* The code requires the musiXmatch dataset to have been loaded into 
an H2 SQL database.
 
* A matrix of word co-occurences is then assembled, with element (i,j) 
the number of tracks in which words (i) and (j) appear together
(standardised by the total number of tracks in which word (i) appears). 
Each word is hence represented in the matrix by the vector of its 
"co-words" in the dataset. The rationale behind this is that if two 
words have the same set of co-words, then these two words are likely
 to belong to the same language.

* A hierarchical linkage of the words is performed in Matlab using 
cosine distance and unweighted average to compute the distance between clusters. 
The resulting dendrogram was sliced at different levels to produce clusterings 
with between two and thirty clusters. 

* To help decide how many clusters to consider, each cluster is submitted to 
the [language-detection](https://code.google.com/archive/p/language-detection/)
Java library. For each cluster, the detection is 
performed over a thousand random phrases of fifty words. Each cluster is 
allocated the language that seemed the most likely.

* Each cluster is also attributed a measurement of "compactness", 
calculated as the cosine of the angle between the cluster and the set 
of words engendered by the cluster. The higher this cosine, the least the 
cluster correlates with other words in the dataset. 

* Finally, the database is queried again in order to produce a sample list
 of artists for every cluster (naively: top x artists that use the top y words 
 in the cluster)

The output of this computation is an [edn](https://github.com/edn-format/edn) file. 
A Jupyter notebook is then used to view the output and assess the optimal number of clusters. 

## Requirements

The analysis requires:
 
* [Leiningen](https://leiningen.org/), the build automation tool for Clojure.
 
* The [H2 database of MSD lyrics](https://github.com/belovehq/msd-lyrics-to-h2), whose path
needs to be edited in `clustering.clj`.

* [Matlab](https://github.com/belovehq/msd-lyrics-to-h2), the numerical computing environment by Mathworks. 
The Matlab Java engine API should be added to the local Leiningen repository with the `lein-localrepo` plugin.

* [Jupyter notebook](http://jupyter.org/), together with the [BeakerX](http://beakerx.com/) collection of Jupyter kernels. 

## Usage

To run the computation with Leiningen: 

* Edit the dependency to Matlab in `project.clj` to match the version of the Matlab engine 
registered in your local Leiningen repo. 

* Edit the path to the database in `clustering.clj`. 

* Then type `lein run` to run the analysis with the parameters 
stored in `resources/clustering/default.edn`. See the code of `clustering.clj` 
for the meaning of parameters.

* Alternatively, the computation can be run from the REPL, 
e.g in namespace `msdlang.clustering`:

```clojure
(def out 
  (run {:requery   true
        :matlab    {:max-clusters 30 :average "average" :distance "cosine"}
        :language  {:phrase-size 50 :phrase-count 1000}
        :wholeness {:limit 200}
        :artists   {:artist-count 10 :word-count 20}
        :out       "out.edn"}))
``` 

Whereas the cluster analysis itself only takes a short time, SQL queries in H2
take a long time (close to 1h on the machine on which this project was coded). 
This includes creating the cowords matrix (pre-analysis) and 
then requesting the artists that use words in clusters (post-analysis). To speed things up: 

* The code caches the cowords matrix in `/resources/clustering/data/cowords.edn`, 
so the cost of building the matrix only has to be paid once. Set the `:requery`
parameter to `true` to query the database a first time, 
and to `false` subsequently to use the cached file.

* The request of the top artists per cluster can be turned off by removing `:artists` 
from the parameters map. 

Once the analysis has run, launch `jupyter notebook` and open the notebook
in `resources/clustering/jupyter` to explore the results stored in 
`/resources/clustering/data/out.edn`. See the 
[online copy](https://nbviewer.jupyter.org/gist/duchenne/77b54f6adb7783848edb9a6e97880c64)
of the notebook.

## License


Copyright © 2018 Nicolas Duchenne, Belove Ltd, London, UK

Released under the [MIT License](https://opensource.org/licenses/MIT).

Based on the musiXmatch dataset, the official lyrics collection for the Million Song Dataset, 
             available at: http://labrosa.ee.columbia.edu/millionsong/musixmatch


CHANGELOG.md
# Change Log


## 0.1.1-SNAPSHOT - 2018-07-21
### Added
- cluster analysis
- Jupyter notebook
- Readme file


src/msdlang/clustering.clj
(ns msdlang.clustering
  (:require
    [clojure.java.jdbc :as j]
    [clojure.java.io :refer [reader writer file resource as-file]]
    [cld.core :as cld]
    [hugsql.core :as h])
  (:import [com.mathworks.engine MatlabEngine]))


; initialisation of language detection

(cld/default-init!)


; parameters

(def paths {:params "clustering"
            :data   "clustering/data"
            :matlab "clustering/matlab"})

(def db {:classname        "org.h2.Driver"
         :subprotocol      "h2"
         :subname          "~/data/millionsong/dataset/h2/msd"
         :user             ""
         :password         ""
         :AUTO_SERVER      "TRUE"
         :AUTO_SERVER_PORT "9091"
         :DB_CLOSE_DELAY   "-1"})

(def languages
  {:af "Afrikaans", :ar "Arabic", :bg "Bulgarian", :bn "Bengali", :cs "Czech",
   :da "Danish", :de "German", :el "Greek", :en "English", :es "Spanish",
   :et "Estonian", :fa "Persian", :fi "Finnish", :fr "French", :gu "Gujarati",
   :he "Hebrew", :hi "Hindi", :hr "Croatian", :hu "Hungarian", :id "Indonesian",
   :it "Italian", :ja "Japanese", :kn "Kannada", :ko "Korean", :lt "Lithuanian",
   :lv "Latvian", :mk "Macedonian", :ml "Malayalam", :mr "Marathi", :ne "Nepali",
   :nl "Dutch", :no "Norwegian", :pa "Punjabi", :pl "Polish", :pt "Portuguese",
   :ro "Romanian", :ru "Russian", :sk "Slovak", :sl "Slovene", :so "Somali",
   :sq "Albanian", :sv "Swedish", :sw "Swahili", :ta "Tamil", :te "Telugu",
   :th "Thai", :tl "Tagalog", :tr "Turkish", :uk "Ukrainian", :ur "Urdu",
   :vi "Vietnamese", :zh-cn "Simplified Chinese", :zh-tw "Traditional Chinese"})


; database queries

(h/def-db-fns "clustering/sql/h2.sql")

(defn query-words [conn]
  (rest (h2-select-words conn {} {} {:as-arrays? true})))

(defn query-words-stems [conn]
  (h2-select-words-stems conn))

(defn query-cowords [conn wordid]
  (rest (h2-select-cowords conn {:wordid wordid} {} {:as-arrays? true})))

(defn query-artists [conn top wordids]
  (rest (h2-select-artists conn {:top top :wordids wordids} {} {:as-arrays? true})))


; utility functions

(defn fullname
  "returns the fullname of a file based on path key and file name
   e.g. (fullname :matlab) or (fullname :matlab \"test.txt\""
  ([k]
   (->> k
        (get paths)
        resource
        as-file
        str))
  ([k f] (str (fullname k) (java.io.File/separator) f)))

(defn ms-to-hms
  "converts a number of milliseconds to hh:mm:ss format"
  [ms]
  (.format
    (.plusSeconds java.time.LocalTime/MIN (/ ms 1000.0))
    java.time.format.DateTimeFormatter/ISO_LOCAL_TIME))

(defn update-map
  "apply function f to all values of map m"
  [f m]
  (reduce-kv (fn [mm k v] (assoc mm k (f v))) {} m))

(defn str-to-double [x]
  "convert a string to double"
  (Double/parseDouble x))


; cluster analysis

(defn get-cowords
  "Return the list of words and their co-occuring words.
   requery: if true then query the database and cache the result into cowords.edn (38 minutes).
   if false then just read the data from cowords.edn."
  ([]
   (get-cowords false))
  ([requery]
   (let [f (fullname :data "cowords.edn")]
     (if requery
       (with-open [out (clojure.java.io/writer f :encoding "UTF-8")]
         (.write out "[")
         (j/with-db-connection
           [conn db]
           (let [words (apply assoc (sorted-map) (flatten (query-words conn)))
                 get-cw (fn [wordid] [wordid
                                      (get words wordid)
                                      (vec (query-cowords conn wordid))])]
             (dorun (map #(.write out (str (get-cw %) \newline))
                         (keys words)))
             (.write out "]")))))
     (read-string (slurp f :encoding "UTF-8")))))


(defn coword-matrix
  "converts the list of words/cowords to a dense double[][]"
  [cowords]
  (let [zeros (vec (repeat (count cowords) 0))
        matrix-row (fn [[_ _ cw]]
                     (reduce
                       #(assoc %1 (dec (first %2)) (second %2))
                       zeros
                       cw))]
    (into-array (map (comp double-array matrix-row) cowords))))


(defn matlab-write-files
  "writes the list of words and the matrix of words/cowords to files for use in matlab."
  []
  (let [cowords (get-cowords)
        write-matrix (fn [f m]
                       (spit f (->> m
                                    (map #(apply str (interpose \space %)))
                                    (interpose \newline)
                                    (apply str))
                             :encoding "UTF-8"))]
    (write-matrix (fullname :matlab "cowords.txt") (coword-matrix cowords))
    (write-matrix (fullname :matlab "words.txt") (map (partial take 2) cowords))))


(defn matlab-cluster
  "perform a cluster analysis of the words/cowords matrix in Matlab.
   cowords: list of words/cowords
   max-clusters: max number of clusters to return data for. Word groupings will be returned for 2 ... max-clusters clusters
   average: the average function to use in Matlab, e.g. 'average'
   distance: the distance function to use in Matlab, e.g. 'cosine'
   params: alternatively, a map of the above parameters {:max-clusters 30 :average \"average\" :distance \"cosine\"}"
  ([cowords params]
   (matlab-cluster cowords (:max-clusters params) (:average params) (:distance params)))
  ([cowords max-clusters average distance]
   (let [engine (MatlabEngine/startMatlab)]
     (try
       (do
         (.eval engine (str "cd " (fullname :matlab)))
         (.eval engine "set(0,'DefaultFigureVisible','off');")
         (.feval engine "clusterwords"
                 (to-array [(coword-matrix cowords)
                            (double max-clusters)
                            average
                            distance])))
       (catch Exception e (println e))
       (finally (.close engine))))))


(defn detect-language
  "Detect the language of a list of words. The detection is performed several times on random phrases, drawn
  in a way that maximises the coverage of the list. The result is the average of the detections over all the phrases.
   phrase-size: the size of the phrases to perform the detection upon.
   phrase-count: how many phrases to perform the detection upon.
   words: the list of words"
  ([params words]
   (detect-language (:phrase-size params) (:phrase-count params) words))
  ([phrase-size phrase-count words]
   (let [; infinite seq of random phrases of given size
         phrases (mapcat #(partition (min phrase-size (count words)) (shuffle %)) (repeat words))

         ; language-detection phrase by phrase
         phrases-lang (map #(->> %
                                 (interpose \space)
                                 (apply str)
                                 cld/detect
                                 second
                                 (update-map str-to-double))
                           (take phrase-count phrases))]

     ; average the language detections across all phrases and sort from most likely to least likely
     (->> phrases-lang
          (apply merge-with +)
          (reduce-kv (fn [m k v] (assoc m (keyword k) (/ v phrase-count))) {})
          (sort-by (comp - second))))))


(defn language-stats [clusters low]
  "Language stats for one clustering.
   Returns a map with:
   :wavg : weighted average or language probabilities;
   :avg : mean of language probabilities over clusters that have more than one word;
   :=1 : count of clusters that have only one word in them."
  (let [obs (fn [{c :count prop :proportion {proba :proba} :language}]
              {:clusters 1
               :wavg     (* prop proba)
               :avg      (if (> c 1) proba 0)
               :=1       (if (> c 1) 0 1)
               :dupe     0})
        stats (->> clusters
                   (map obs)
                   (apply merge-with +))
        langs>1 (->> clusters
                     (filter #(> (:count %) 1))
                     (map #(get-in % [:language :id])))]
    (-> stats
        (update :avg / (- (double (count clusters)) (:=1 stats)))
        (assoc :dupe (- (count langs>1) (count (distinct langs>1)))))))


(defn get-artists
  "return a sample list of artists who use words in  a cluster.
  dbconn: a database connection or spec
  cluster: cluster data, see code of function analyze-clustering further down
  artist-count: how many artists to return
  word-count: how many words to take in the cluster to perform artist detection"
  ([dbconn params cluster]
   (if params
     (get-artists dbconn (:artist-count params) (:word-count params) cluster)
     []))
  ([dbconn artist-count word-count cluster]
   (flatten (query-artists dbconn artist-count (take word-count (:wordids cluster))))))


(defn compactness [wordids cowords params]
  "Calculate the compactness of a cluster as the cosine between the cluster itself and the
   words engendered by the cluster"
  (let [wids (set wordids)
        c (double (count wids))
        generated (->> cowords
                       (filter (comp wids first))
                       shuffle
                       (take (:limit params))
                       (map #(apply hash-map (flatten (last %))))
                       (reduce #(merge-with + %1 %2))
                       (update-map #(/ % c))
                       (into (sorted-map)))
        generator (select-keys generated wids)
        n2 (fn [m] (->> m
                        vals
                        (reduce #(+ %1 (* %2 %2)) 0)
                        double))]
    ;by construction here: cosine = |generator| / |generated|
    {:generated generated
     :cosine    (Math/sqrt (/ (n2 generator) (n2 generated)))}))


(defn analyze-clustering
  [cowords wordsmap lang-params wholeness-params matlab-clustering]
  " Analyze a clustering produced by matlab.
  wordsmap: the map of wordids to full words
  lang-params: map of parameters for the language recognition, see function detect-language
  artist-params: map of parameters for the artist report, see function get-artists
  matlab-clustering: the vector of clusterids returned by matlab."
  (let [;converts a Matlab clustering to a list of maps
        base-clusters (->> matlab-clustering
                           (map int)
                           (interleave (rest (range)))
                           (partition 2)
                           (group-by second)
                           (map (fn [[k v]] (assoc {} :clusterid k :wordids (sort (map first v)))))
                           (sort-by :clusterid))

        ;analyze a single cluster
        analyze-cluster (fn [cluster]
                          (let [words (map wordsmap (:wordids cluster))
                                [[langid langpr] & _ :as langs] (detect-language lang-params words)
                                {:keys [cosine generated]} (compactness (:wordids cluster) cowords wholeness-params)]
                            (assoc cluster
                              :words words
                              :count (count words)
                              :proportion (/ (double (count words)) (count matlab-clustering))
                              :language {:id langid :proba langpr :detect langs}
                              :generated generated
                              :compactness cosine)))

        ;analyse all clusters
        clusters (map analyze-cluster base-clusters)

        ; print cluster count
        _ (println (count clusters) "clusters")]

    {:count          (count clusters)
     :matlab         (map int matlab-clustering)
     :clusters       (vec clusters)
     :language-stats (language-stats clusters (:low lang-params))}))



(defn get-all-artists [artists-param clusterings]
  (if artists-param

    (let [cluster-artists (fn [dbconn c] (assoc c :artists (get-artists dbconn artists-param c)))
          do-clusters (fn [dbconn clusters] (vec (map (partial cluster-artists dbconn) clusters)))]
      (j/with-db-connection
        [conn db]
        (->> clusterings
             (map #(update % :clusters (partial do-clusters conn)))
             vec
             doall)))

    clusterings))

; Run a whole analysis

(defn run
  ([params] (run params (:requery params)))

  ([params requery]
   (let [t0 (System/currentTimeMillis)

         ; retrieve the cowords matrix
         cowords (get-cowords requery)
         words-map (apply array-map (flatten (map (partial take 2) cowords)))

         ; run the matlab clustering
         {:keys [max-clusters average distance]} (:matlab params)
         matlab-clusterings (matlab-cluster cowords max-clusters average distance)

         ; analyse the matlab clustering
         clusterings (->> matlab-clusterings
                          (pmap (partial analyze-clustering
                                         cowords
                                         words-map
                                         (:language params)
                                         ;(:artists params)
                                         (:wholeness params)))
                          doall
                          (get-all-artists (:artists params)))

         ; retrieve word stems as map indexed on wordid
         stems (reduce #(assoc %1 (:wordid %2) (dissoc %2 :wordid))
                       (sorted-map)
                       (h2-select-words-stems db))

         ; output results
         out {:params      params
              :languages   languages
              :words       stems
              :clusterings clusterings}

         _ (spit (fullname :data (:out params)) out :encoding "UTF-8")
         _ (println "Run time:" (ms-to-hms (- (System/currentTimeMillis) t0)))]

     out)))


; run with Leiningen

(defn -main
  "Runs the default cluster analysis, as per resources/clustering/default.edn"
  [& args]
  (->> "default.edn"
       (fullname :params)
       slurp
       read-string
       run
       dorun))


; REPL candy

(comment
  (def out (run {:requery   false
                 :matlab    {:max-clusters 30 :average "average" :distance "cosine"}
                 :language  {:phrase-size 50 :phrase-count 1000}
                 :wholeness {:limit 200}
                 :artists   {:artist-count 10 :word-count 20}
                 :out       "out.edn"})))


